File: 03. Stack and Queue/05. Previous Greater Element Circular/main.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> previousGreaterElementCircular(vector<int>& nums) {
    int n = nums.size();
    vector<int> result(n, -1);
    stack<int> st;

    // Traverse the array twice in reverse order to handle circular behavior
    for (int i = 2 * n - 1; i >= 0; i--) {
        while (!st.empty() && st.top() <= nums[i % n]) {
            st.pop();
        }
        if (i < n) {
            if (!st.empty()) {
                result[i] = st.top();
            }
            st.push(nums[i]);
        }
    }

    return result;
}

int main() {
    vector<int> nums = {4, 5, 2, 10, 8};
    vector<int> result = previousGreaterElementCircular(nums);

    cout << "Previous Greater Elements (Circular):\n";
    for (int num : result) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}

File: 03. Stack and Queue/05. Previous Greater Element Circular/main.java
import java.util.*;

public class main {
    public static int[] previousGreaterElements(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        Arrays.fill(result, -1);
        Stack<Integer> stack = new Stack<>();

        // Traverse twice in reverse order to simulate circular array
        for (int i = 2 * n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && stack.peek() <= nums[i % n]) {
                stack.pop();
            }
            if (i < n) {
                if (!stack.isEmpty()) {
                    result[i] = stack.peek();
                }
                stack.push(nums[i]);
            }
        }

        return result;
    }

    public static void main(String[] args) {
        int[] nums = {4, 5, 2, 10, 8};
        int[] result = previousGreaterElements(nums);

        System.out.println("Previous Greater Elements (Circular):");
        System.out.println(Arrays.toString(result));
    }
}
